neonvm/tools/vm-builder-generic/main.go

```diff
@@ -298,98 +308,40 @@ func main() {
 		log.Fatalln(err)
 	}
 
-	var SrcImageSpec ImageSpec
+	tmplCtx := TemplatesContext{
+		SrcImage:   *srcImage,
+		Entrypoint: imageSpec.Config.Entrypoint,
+		Cmd:        imageSpec.Config.Cmd,
+		Env:        imageSpec.Config.Env,
+	}
 
 	if len(imageSpec.Config.User) != 0 {
-		SrcImageSpec.User = imageSpec.Config.User
+		tmplCtx.User = imageSpec.Config.User
 	} else {
-		SrcImageSpec.User = "root"
+		tmplCtx.User = "root"
 	}
 
-	SrcImageSpec.Entrypoint = imageSpec.Config.Entrypoint
-	SrcImageSpec.Cmd = imageSpec.Config.Cmd
 	// if no entrypoint and cmd in docker image then use sleep for 10 years as stub
-	if len(SrcImageSpec.Entrypoint) == 0 && len(SrcImageSpec.Cmd) == 0 {
-		SrcImageSpec.Cmd = []string{"/neonvm/bin/sleep", "3650d"}
+	if len(tmplCtx.Entrypoint) == 0 && len(tmplCtx.Cmd) == 0 {
+		tmplCtx.Cmd = []string{"/neonvm/bin/sleep", "3650d"}
 	}
 
-	SrcImageSpec.Env = imageSpec.Config.Env
-
 	buf := new(bytes.Buffer)
 	tw := tar.NewWriter(buf)
 	defer tw.Close()
 
-	// generate Dockerfile from template
-	dockerfileVmBuilderTmpl, err := template.New("vm-builder").Parse(dockerfileVmBuilder)
-	if err != nil {
-		log.Fatal(err)
-	}
-	var dockerfileVmBuilderBuffer bytes.Buffer
-	err = dockerfileVmBuilderTmpl.Execute(&dockerfileVmBuilderBuffer, *srcImage)
-	if err != nil {
-		log.Fatalln(err)
-	}
-
-	// generate vmstart script from template
-	scriptVmStartTmpl, err := template.New("vmstart").Parse(scriptVmStart)
-	if err != nil {
-		log.Fatal(err)
-	}
-	var scriptVmStartBuffer bytes.Buffer
-	err = scriptVmStartTmpl.Execute(&scriptVmStartBuffer, SrcImageSpec)
-	if err != nil {
-		log.Fatalln(err)
-	}
-
-	// add 'Dockerfile' file to docker build context
-	if err = AddToTar(tw, "Dockerfile", dockerfileVmBuilderBuffer); err != nil {
-		log.Fatalln(err)
-	}
-	// add 'vmstart' file to docker build context
-	if err = AddToTar(tw, "vmstart", scriptVmStartBuffer); err != nil {
-		log.Fatalln(err)
-	}
-
-	var b bytes.Buffer
-
-	// add 'inittab' file to docker build context
-	_, err = b.WriteString(scriptInitTab)
-	if err != nil {
-		log.Fatalln(err)
-	}
-	if err = AddToTar(tw, "inittab", b); err != nil {
-		log.Fatalln(err)
-	}
-
-	// add 'vmacpi' file to docker build context
-	b.Reset()
-	_, err = b.WriteString(scriptVmAcpi)
-	if err != nil {
-		log.Fatalln(err)
-	}
-	if err = AddToTar(tw, "vmacpi", b); err != nil {
-		log.Fatalln(err)
-	}
-
-	// add 'powerdown' file to docker build context
-	b.Reset()
-	_, err = b.WriteString(scriptPowerDown)
-	if err != nil {
-		log.Fatalln(err)
-	}
-	if err = AddToTar(tw, "powerdown", b); err != nil {
-		log.Fatalln(err)
+	add := func(filename string, tmpl string) {
+		if err := AddTemplatedFileToTar(tw, buf, tmplCtx, filename, tmpl); err != nil {
+			log.Fatalln(err)
+		}
```

Comment:
I'd just write a loop rather than do this functional thing.

Reply:
Makes sense. Fixed